goftests/utils.py

```diff
@@ -0,0 +1,100 @@
+#   may be used to endorse or promote products derived from this
+#   software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+# TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+This module computes the chi^2 survival function and the required
+functions.
+
+This was implemented so that we could use the goftests module
+with pypy. Given that the only scipy dependence inside goftests
+was from the chi^2 survival function (1 - cdf), and then gamma
+function.
+
+"""
+
+from __future__ import division
+import math
+import sys
+from numpy import inf
+
+
+def log(x):
+    if x > sys.float_info.min:
+        value = math.log(x)
+    else:
+        value = -inf
+    return value
+
+
+def incomplete_gamma(x, s):
+    r"""
+    This function computes the incomplete lower gamma function
+    using the series expansion:
+    \gamma(x, s) = x^s \Gamma(s)e^{-x}\sum^\infty_{k=0}
+                    \frac{x^k}{\Gamma(s + k + 1)}
+    This series will converge strongly because the Gamma
+    function grows factorially.
+
+    Because the Gamma function does grow so quickly, we can
+    run into numerical stability issues. I have chosen to carry
+    out as much math as possible in the log domain to reduce
+    numerical error. This function matches the results from
+    scipy to numerical precision.
+    """
+    value = 0.0
+    if x < 0.0:
+        return 1.0
+    if x > 1e3:
+        return math.gamma(s)
+    log_gamma_s = log(math.gamma(s))
+    log_x = log(x)
+    for k in range(0, 100):
+        log_num = (k + s)*log_x + (-x) + log_gamma_s
+        log_denom = log(math.gamma(k + s + 1))
+        value += math.exp(log_num - log_denom)
+    return value
+
+
+def chi2sf(x, s):
+    r"""
+    This function returns the survival function of the chi^2
+    distribution. The survival function is given as:
+    1 - CDF
+    where rhe chi^2 CDF is given as
+    F(x; s) = \frac{ \gamma( x/2, s/2 ) }{ \Gamma(s/2) },
+    with $\gamma$ is the incomplete gamma function defined above.
+    Therefore, the survival probability is givne by:
+    1 - \frac{ \gamma( x/2, s/2 ) }{ \Gamma(s/2) }.
+    This function matches the results from
+    scipy to numerical precision.
+    """
+    top = incomplete_gamma(x/2.0, s/2.0)
```

Comment:
The `2.0`s and `1.0`s are all unnecessary as long as you have `from __future__ import division`; it's more readable to just do `x / 2`, etc.